ovirt_imageio/tool/ovirt.py

```diff
@@ -0,0 +1,355 @@
+
+    Using the local host for an image transfer allows optimizing the
+    connection using unix socket. This speeds up the transfer
+    significantly and minimizes the network bandwidth.
+
+    However using the local host is possible only if:
+    - The local host is a oVirt host
+    - The host is Up
+    - The host is in the same DC of the storage domain
+
+    Consider this setup:
+
+        laptop1
+
+        dc1
+            host1 (down)
+            host2 (up)
+            sd1
+                disk1
+
+        dc2
+            host3 (up)
+            sd2
+                disk2
+
+    - If we run on laptop1 we cannot use the local host for any
+      transfer.
+    - If we run on host1, we cannot use the local host for any transfer.
+    - If we run on host2, we can use use host2 for transferring disk1.
+    - If we run on host3, we can use host3 for transferring disk2.
+
+    Arguments:
+        con (ovirtsdk4.Connection): Connection to ovirt engine
+        sd_name (str): Storage domain name
+
+    Returns:
+        ovirtsdk4.types.Host
+    """
+
+    # Try to read this host hardware id.
+
+    try:
+        with open("/etc/vdsm/vdsm.id") as f:
+            vdsm_id = f.readline().strip()
+    except FileNotFoundError:
+        log.debug("Not running on oVirt host, using any host")
+        return None
+    except OSError as e:
+        # Unexpected error when running on ovirt host. Since choosing a
+        # host is an optimization, log and continue.
+        log.warning("Cannot read /etc/vdsm/vdsm.id, using any host: %s", e)
+        return None
+
+    log.debug("Found host hardware id: %s", vdsm_id)
+
+    # Find the data center by storage domain name.
+
+    system_service = con.system_service()
+    data_centers = system_service.data_centers_service().list(
+        search=f'storage.name={sd_name}',
+        case_sensitive=True)
+    if len(data_centers) == 0:
+        raise RuntimeError(
+            f"Storage domain {sd_name} is not attached to a DC")
+
+    data_center = data_centers[0]
+    log.debug("Found data center: %s", data_center.name)
+
+    # Validate that this host is up and in data center.
+
+    hosts_service = system_service.hosts_service()
+    query = f"hw_id={vdsm_id} and datacenter={data_center.name} and status=Up"
+    hosts = hosts_service.list(search=query, case_sensitive=True)
+    if len(hosts) == 0:
+        log.debug(
+            "Cannot use host with hardware id %s, host is not up, or does "
+            "not belong to data center %s",
+            vdsm_id, data_center.name)
+        return None
+
+    host = hosts[0]
+    log.debug("Using host id %s", host.id)
+
+    return host
+
+
+def create_transfer(
+        con, disk=None, direction=types.ImageTransferDirection.UPLOAD,
+        host=None, backup=None, inactivity_timeout=None, timeout=60,
+        disk_snapshot=None, shallow=None,
+        timeout_policy=types.ImageTransferTimeoutPolicy.CANCEL):
```

Comment:
This is ok-ish, but when I see so many arguments in a function, where some of them are mutually exclusive (e.g., `disk` and `disk_snapshot`), I feel this could be a good place for data classes with inheritance.

This way we would receive only a single argument, the base class (polymorphism). And we will avoid having most of the if elses inside the function, as the data class will give us the correct ID, transfer, etc.

Reply:
Only disk and disk_snapshot are  mutually exclusive, and this is unfortunately the underlying engine API we have to work with.

I don't see how data classes can help, and they are not available in python 3.6 so
we cannot use them anyway.

In the original code when we added support for disk snapshots we did this in 
a backward compatible way so we don't break users using the existing helper
scripts, but we don't have this issue now since nobody is depending on this
and this is not a library but internal module.

How about using:

    create_tranfer(con, source=disk)
    create_tranfer(con, source=disk_snapshot)

And internally handle the type of the source?

    if isinstance(source, types.DiskSnapshot):
        transfer.snapshot = source
    elif isinstance(source, types.Disk):
        transfer.disk = source
    else:
        raise ValueError(...)

Reply:
I mentioned dataclasses, but it can be solved with plain classes too (main difference is the explicit constructor anyway).

But nevermind that, what you proposed looks good too :)